server/src-lib/Hasura/GraphQL/Resolve/Mutation.hs

```diff
@@ -115,53 +118,67 @@ convertUpdateP1
-  mutFlds <- convertMutResp (_fType fld) $ _fSelSet fld
-
-  let resolvedPreSetItems =
-        Map.toList $ fmap partialSQLExpToUnresolvedVal preSetCols
+  deleteAtPathExpM <- resolveUpdateOperator "_delete_at_path" $
+    convDeleteAtPathObj colGNameMap
 
-  let updExpsM = [ setExpM, incExpM, appendExpM, prependExpM
+  updateItems <- combineUpdateExpressions
+                 [ setExpM, incExpM, appendExpM, prependExpM
                  , deleteKeyExpM, deleteElemExpM, deleteAtPathExpM
                  ]
-      setItems = resolvedPreSetItems ++ concat (catMaybes updExpsM)
-
-  -- atleast one of update operators is expected
-  -- or preSetItems shouldn't be empty
-  -- this is not equivalent to (null setItems)
-  unless (any isJust updExpsM || not (null resolvedPreSetItems)) $ throwVE $
-    "atleast any one of _set, _inc, _append, _prepend, "
-    <> "_delete_key, _delete_elem and "
-    <> "_delete_at_path operator is expected"
 
-  let unresolvedPermFltr = fmapAnnBoolExp partialSQLExpToUnresolvedVal filterExp
+  mutFlds <- convertMutResp (_fType fld) $ _fSelSet fld
 
-  return $ RU.AnnUpd tn setItems
-    (unresolvedPermFltr, whereExp) mutFlds allCols
+  pure $ RU.AnnUpd tn updateItems (unresolvedPermFilter, whereExp) mutFlds allCols
   where
     convObjWithOp' = convObjWithOp colGNameMap
     allCols = Map.elems colGNameMap
     UpdOpCtx tn _ colGNameMap filterExp preSetCols = opCtx
     args = _fArguments fld
+    resolvedPreSetItems = Map.toList $ fmap partialSQLExpToUnresolvedVal preSetCols
+    unresolvedPermFilter = fmapAnnBoolExp partialSQLExpToUnresolvedVal filterExp
+
+    resolveUpdateOperator operator resolveAction =
+      (operator,) <$> withArgM args operator resolveAction
+
+    combineUpdateExpressions updateExps = do
+      let allOperatorNames = map fst updateExps
+          updateItems = mapMaybe (\(op, itemsM) -> (op,) <$> itemsM) updateExps
+      -- Atleast any one of operator is expected or preset expressions shouldn't be empty
+      if null updateItems && null resolvedPreSetItems then
+        throwVE $ "atleast any one of " <> showNames allOperatorNames <> " is expected"
+      else do
+        let itemsWithOps = concatMap (\(op, items) -> map (second (op,)) items) updateItems
+            itemTuples = OMap.toList $ OMap.groupTuples itemsWithOps
+            columnsWithMultiOps = map (second (map fst . NESeq.toList)) $
+                                  filter ((>1) . NESeq.length . snd) itemTuples
+        -- A column shouldn't be present in more than one operator.
+        -- If present, then generated UPDATE statement throws unexpected query error
+        if null columnsWithMultiOps then
+          pure $ resolvedPreSetItems <> map (second (snd . NESeq.head)) itemTuples
+        else throwVE $ "column found in multiple operators; "
+                       <> T.intercalate ". "
+                          (map (\(col, ops) -> col <<> " in " <> showNames ops)
+                               columnsWithMultiOps)
```

Comment:
This generally looks good to me, but one suggestion I have to make this a little more data-/type-driven would be to take advantage of the result of `OMap.groupTuples` more aggressively. That operation produces a map with a shape like `InsOrdHashMap k (NESeq v)`, and from there you can very naturally write a function of type

```haskell
(MonadError QErr m) => k -> NESeq v -> m v
```

which raises an error if there are multiple elements and returns a single element otherwise. Then you can hand that function to `OMap.traverseWithKey` to get an `OMap k v` out, which is, in effect, the “parsed” representation of that input. 🙂

Reply:
Yes, I can do `(MonadError QErr m) => k -> NESeq v -> m v`, but it'd stop at the occurrence of first multiple elements. My point is to collect all those multiple elements (`columnsWithMultipleOps`) and throw an error including all of them in the error message.